src/Servers/Kestrel/Transport.Sockets/src/Internal/IOQueue.cs

```diff
@@ -43,14 +45,25 @@ void IThreadPoolWorkItem.Execute()
                     item.Callback(item.State);
                 }
 
-                lock (_workSync)
+                // All work done.
+                // Set _doingWork (0 == false) prior to checking IsEmpty to catch any missed work in interim.
+                _doingWork = 0;
+
+                // Check if there is work to do
+                if (_workItems.IsEmpty)
```

Comment:
I think @Daniel-Svensson is right.  I don't see anything that would prevent the IsEmpty from being observed as occurring before the write to _doWork, even on x86.

Reply:
Dropped `volatile` added `Thread.MemoryBarrier` here and `Volatile.Read` in the other path https://github.com/aspnet/AspNetCore/pull/6154/commits/baeeb243520c81757ddbb9cd368450a7bb0772d9

Reply:
Good catch. Do you think it was the the volatile write to `_doingWork` followed immediately by the volatile read of `_head` in the `ConcurrentQueue.IsEmpty` getter made this technically correct before by effectively combining to create a full memory barrier? Not that I would recommend relying `ConcurrentQueue.IsEmpty` doing a volatile read.

Reply:
> Do you think it was the the volatile write to _doingWork followed immediately by the volatile read of _head in the ConcurrentQueue.IsEmpty getter made this technically correct before by effectively combining to create a full memory barrier?

I don't believe it was technically correct before.  The volatile reads could have moved up above the volatile writes, so in theory this could end up having essentially been the equivalent of:
```C#
bool wasEmpty = _workItems.IsEmpty;
_doingWork = 0;
if (wasEmpty) break;
```
and with that ordering, you could end up with no threads processing work even when there was work in the queue:
- Thread 1 is running Execute processing the last work item, so _doingWork is 1 and the queue is empty.
- Thread 1 finishes processing the last work item and checks IsEmpty, storing true.
- Thread 2 comes along and Enqueue's a work item, then sees that _doingWork is 1 and exits.
- Thread 1 sets _doingWork to 0 and then having seen that IsEmpty was true, breaks out of the loop and exits.

Reply:
Could similar happen on the queue side? e.g. should always execute Exchange and drop the Volatile.Read or add a barrier before the read?

Reply:
> should always execute Exchange and drop the Volatile.Read

Probably.

Reply:
Added a barrier as it frequently already doing work (i.e. is often a read and write skip)